pkg/hubble/parser/sock/parser.go

```diff
@@ -0,0 +1,221 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Hubble
+
+package sock
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+
+	flowpb "github.com/cilium/cilium/api/v1/flow"
+	"github.com/cilium/cilium/pkg/byteorder"
+	"github.com/cilium/cilium/pkg/hubble/parser/common"
+	"github.com/cilium/cilium/pkg/hubble/parser/errors"
+	"github.com/cilium/cilium/pkg/hubble/parser/getters"
+	"github.com/cilium/cilium/pkg/monitor"
+	monitorAPI "github.com/cilium/cilium/pkg/monitor/api"
+)
+
+// Parser is a parser for SockTraceNotify payloads
+type Parser struct {
+	log            logrus.FieldLogger
+	endpointGetter getters.EndpointGetter
+	identityGetter getters.IdentityGetter
+	dnsGetter      getters.DNSGetter
+	ipGetter       getters.IPGetter
+	serviceGetter  getters.ServiceGetter
+	cgroupGetter   getters.PodMetadataGetter
+	epResolver     *common.EndpointResolver
+}
+
+// New creates a new parser
+func New(log logrus.FieldLogger,
+	endpointGetter getters.EndpointGetter,
+	identityGetter getters.IdentityGetter,
+	dnsGetter getters.DNSGetter,
+	ipGetter getters.IPGetter,
+	serviceGetter getters.ServiceGetter,
+	cgroupGetter getters.PodMetadataGetter,
+) (*Parser, error) {
+	return &Parser{
+		log:            log,
+		endpointGetter: endpointGetter,
+		identityGetter: identityGetter,
+		dnsGetter:      dnsGetter,
+		ipGetter:       ipGetter,
+		serviceGetter:  serviceGetter,
+		cgroupGetter:   cgroupGetter,
+		epResolver:     common.NewEndpointResolver(log, endpointGetter, identityGetter, ipGetter),
+	}, nil
+}
+
+// Decode takes a raw trace sock event payload obtained from the perf event ring
+// buffer and decodes it into a flow
+func (p *Parser) Decode(data []byte, decoded *flowpb.Flow) error {
+	if len(data) == 0 {
+		return errors.ErrEmptyData
+	}
+
+	eventType := data[0]
+	if eventType != monitorAPI.MessageTypeTraceSock {
+		return errors.NewErrInvalidType(eventType)
+	}
+
+	sock := &monitor.TraceSockNotify{}
+	if err := binary.Read(bytes.NewReader(data), byteorder.Native, sock); err != nil {
+		return fmt.Errorf("failed to parse sock trace event: %w", err)
+	}
+
+	isRevNat := decodeRevNat(sock.XlatePoint)
```

Comment:
Yes, your suspicion is correct, an earlier version of the PR used the result of  `decodeRevNat` as the reply flag. I've changed it to a `bool` in the most recent version.

Unfortunately, trying to derive the traffic direction / reply state from the observation point does not work: `IsReply` cannot be meaningfully populated for TraceSock events because their occur on the socket granularity level, not a packet level. 

The screenshot in the PR description only shows the traces from the node where `pod-to-a` is hosted, where rev nat is indeed only used for the reply. But if we include also the events emitted by the kube-dns host, then the picture looks slightly different, we also see rev nat being used for the non-reply packet:

![image](https://user-images.githubusercontent.com/50564/199265945-4f66fa6d-909f-43eb-b6b3-83c2b0c16b34.png)

The screenshot above is using `xwing`, but it's a single UDP request packet being sent from xwing to kubedns, with a single reply packet being sent back. Please note the timetsamps - as you can see, `kind-worker2` (where coredns is hostd) emits a `pre-xlate-rev` for the initial incoming packet (there is no post event, as there is no translation). The reply packet (the last 4 events) happens half a second later, and does not emit any SockLB event, probably because it's sent back on the same "connected UDP socket".

Reply:
Another way to formulate this: I think (but I might be mistaken, cc @aditighag) that `pre/post-xlate-rev` is only executed on `recvmsg` and `getpeername`, both system calls don't really imply anything of overall the "traffic direction" (e.g. if it was an ingress or egress connection in policy lingo).

`pre/post-xlate-fwd` on the other hand is executed for `connect` and `sendmsg`. `connect` is typically used to initiate an egress connection, but the same cannot necessary be said for `sendmsg`.

Reply:
@gandro ah that make sense, thanks for the detailed explanation :+1: one more question about the example:

> The reply packet (the last 4 events) happens half a second later, and does not emit any SockLB event, probably because it's sent back on the same "connected UDP socket".

To make sure I understand correctly: the reply packet does not emit any SockLB event on `kind-worker2` (i.e. `coredns` host), but it does in `kind-worker` (i.e. `xwing` host) correct? The screenshot's last two events are a couple of `pre/post-xlate-rev` analogous to the request packet `pre/post-xlate-fwd` events (which make sense to me), hence the question.

Reply:
That's correct, yes. My guess here is that `coredns` is using Golang's `net.PacketConn` to send the reply packet, meaning it's using the `write` system call on a "connected UDP socket". Thus, none of the above cgroup hooks is hit. `dig` (used inside `xwing`) on the other hand might be using `readmsg`/`sendmsg` syscalls, thus causing a Hubble event for each syscall.

Maybe @aditighag has deeper insight here.

Reply:
On second thought, connected UDP sockets are only useful for clients. Checking a basic UDP echo server, it seems Go is using `recvfrom` and `sendto`, so not sure why the `sendto` does not cause a Hubble event.

Reply:
The recvmsg hook is executed only for regular UDP. TCP (and connected UDP) sockets are connected, so TCP clients send and receive data using `send`/`recv` socket calls.
@gandro Your observation is correct. Golang uses a connected UDP client by default. Regarding, `sendto` for connected UDP, here is a snippet from the man page - 

`If sendto() is used on a connection-mode (SOCK_STREAM, SOCK_SEQPACKET) socket, the arguments dest_addr and addrlen are ignored (and the error EISCONN may be returned when they are not NULL and 0)`

Reply:
>Another way to formulate this: I think (but I might be mistaken, cc @aditighag) that pre/post-xlate-rev is only executed on recvmsg and getpeername, both system calls don't really imply anything of overall the "traffic direction" (e.g. if it was an ingress or egress connection in policy lingo).

When podA on nodeA talks to podB on nodeB, both fwd and rev translation happen on the nodeA (for UDP). For TCP and connected UDP, it would just be fwd translation.  

The {pre, post}-fwd/rev markers already convey the direction. We shouldn't try to interpret this with the same context as policy terminology (ingress/egress).

fwd: svc vip -> backend ip
rev: backend ip -> svc vip

Hope that makes sense.